Contributions/Linux_Kernel_Safety_First_Principles.md

```diff
@@ -0,0 +1,141 @@
+This is not intended to be a complete list of facts, however it attempts
+to summarise the most relevant ones from the documents listed in the
+References section.
+
+## **Purpose of the document**
+Summarise the concepts from the documents in the Reference sections.
+Justifications for each statement can be found throughout those documents.
+
+## **Structure of the document**
+The next section illustrates the philosohpy behind the approach used here to
+deal with safety requirements vs a system using Linux.
+The further two sections about "First Principles" address two aspects of safety:
+-  integrity: reliably detecting interference to components with safety requirements
+-  availability: protracted integrity of the safey-relevant functionality
+The last section lists considerations that can help with raching the safety goals.
+
+
+## **Philosophy of the Document**
+*  A system subject to safety goals might employ the Linux Kernel.
+*  Safety goals are specific to the overall system and the related use cases.
+*  Similarly, one must expect that the solution will be tailored to both system and goals.
+*  However, it is possible to describe the Linux Kernel from a safety point of view.
+*  Not all the issues listed will be necessarily relevant to each system/application.
+*  A Safety Analysis should, however, clearly state which issues are relevant, which are not, and why.
+
+
+## **First Principles - Integrity**
+This section provides touchstone concepts about interference of QM components with
+safety-qualified ones.
+
+1. The vanilla Linux Kernel, is not sufficient to support, alone, safey goals.
+2. No internal kernel barriers/protections against interference to variable data, including safety-relevant data.
+3. Any component within the kernel context can generate (cascaded) interference.
+4. Internal interference can cascade also through components that have been qualified for safety at unit-level.
+5. The Linux Kernel is able to interfere with both itself and any part of user-space processes (linear map).
+6. No generic solution for very complex systematic corruption of any writable memory.
+7. Stress testing alone is not sufficient for supporting safety claims.
+8. Safety claims on the Linux Kernel require both positive testing and focused negative testing.
+9. Those security features which are based on randomisation decrease repeatability of testing (e.g. structure layout randomisation).
+10. Safety claims must be supported by components with the same-or-better safety level (e.g. a safety-qualified watchdog).
+11. Non safety qualified processes can interfere with safety qualified ones, indirectly, through the kernel (e.g. triggering memory management bugs).
+12. Using cgroups/containers/SELinux removes only certain simple types of user-space-induced interference.
+13. In a mixed-criticality scenario, non safety-qualified code represents a safety liability that grows with the frequency of execution (cgroups/containers, LSM/SELinux, etc.).
```

Comment:
Are you saying that the more often NSR code is executed the higher the chance of interference is? If so why are you mentioning cgroups/containers ?

Reply:
NSR = Non Safety Relevant? 
If that is the case, then no, I really mean qualified, as out of context.
cgroups are not qualified - and qualifying them would mean qualifying the entire kernel, more or less
same goes for LSM/SELinux


The fact that one might want to use them to prevent a certain type of interference dpoesn't automagically qualify them :-D

Reply:
@reiterative  I would rephrase as <<In a mixed-criticality scenario, all unqualified code represents a potential risk of interference, which grows with frequency of execution; therefore any code supporting the mixed criticality safety claim (e.g. cgroups/containers, LSM/SELinux, etc.) needs to be qualified according to the highest safety integrity level of the supported safety processes>>

The reason being that the way it is currently written sounds like "cgroups/containers, LSM/SELinux" are bad actors; instead I think we should focus on what needs to be done if we want to leverage such features...

Reply:
But they *ARE* bad actors, since they are not ASIL
Again, evidence supporting the opposite is welcomed.
Till then, I think I stated the existing situation.

Reply:
Then any piece of the Kernel is a bad actor unless you find all the evidence that you need to build a safety case.
IMO the goal of this doc should be to help focusing on the challenges that must be solved for using the Kernel in supporting certain safety claim. So if you need to claim mixed criticality between user space processes you need to focus on the qualification of MM, Scheduler, namespaces, cgroups, SELinux etc. 
Instead if you need to claim FFI between Kernel threads (that also includes interference of a QM driver towards a user space process), you still need to focus on the qualification of MM and scheduler plus you need to model the interference from a QM driver toward an ASIL kernel component and toward a user mode thread.

Reply:
Indeed: there is no reason to expect any safety level from any prt of the kernel, unless proven differently.

The goal of this document is simply to provide *facts* that one can start from.
Speculations can certainly be made available, but not here.